### ISSUE TITLE
Better Solutions

### ISSUE THREAD

**Issue by BasicallyIAmFox** [bluemagic123/tModLoader#3293]
### Do you intend to personally contribute/program this feature?

Yes

### I would like to see this change made to improve my experience with

Mod capability as a Modder

### Description

Adds a new ModType that is called ModConversion. As name says, it allows to add custom conversions. Maybe there also could be GlobalConversion or something like that for modifying existing conversions. This could also add optimized TileScanner class, which as name says, would scan nearby tiles.
Code for custom conversion types would look like that in practice:
```
using Terraria.ID;
using Terraria.ModLoader;
using TheConfectionRebirth.Tiles;
using TheConfectionRebirth.Walls;

public class ConfectionConversion : ModConversion {
    public override void SetStaticDefaults() {
        Conversion
            // Replace moss and all stone convertibles into Creamstone
            .From(Main.tileMoss)
            .From(TileID.Sets.Conversion.Stone)
            .To<Creamstone>()
            .OnConversion(TryKillingTreesAboveIfTheyWouldBecomeInvalid)
            .RegisterTile()
            
            .From(TileID.Sets.Conversion.Ice)
            .To<BlueIce>()
            .OnConversion(TryKillingTreesAboveIfTheyWouldBecomeInvalid)
            .RegisterTile()
            
            .From(TileID.Mud)
            .To<CookieBlock>()
            .BeforeConversion((Tile tile, int i, int j) => {
                if (TileScanner.ScanCircle((tile, i, j) => tile.TileType == ModContent.TileType<CreamGrass>(), i, j, 1)) {
                    return ConversionRunCodeValues.Run;
                }
                // Can't convert if no cream grass nearby!
                return ConversionRunCodeValues.DontRun;
            })
            .RegisterTile()
            
            // Break all thorns
            .From(TileID.Sets.Conversion.Thorn)
            .To(ConversionHandler.Break)
            .RegisterTile()
            
            // Make all grass walls be convertible into cream grass wall
            .From(WallID.Sets.Conversion.Grass)
            .To<CreamGrassWall>()
            .RegisterWall();
    }
}
```
And they would call conversion tile to other tile like that: `WorldGen.Convert<ConfectionConversion>(i, j)` - to convert one tile respectively. Putting a number after coordinates will convert all tiles in X radius using the TileScanner.

### What does this proposal attempt to solve or improve?

Currently every single conversion type is just a static method called "Convert" in their custom solution projectile with generally too much repeative code (just like vanilla, I guess) and barely any readability. This attempts to increase readability as well as simplify all the code used by conversions.

**Comment by BasicallyIAmFox** [comment]
I suppose I can start doing PR?

**Comment by JavidPack** [comment]
This is fairly low priority at this point in the process of updating to 1.4.4, so I haven't given this much thought. You might want to wait for some feedback from other modders who have made conversions, see if they like the syntax you have laid out. To me it appears like a lot of code for such a small feature, there might be a less intrusive way to do it.

**Comment by Chicken-Bones** [comment]
There are higher priority items for the maintenance team at the moment, and these new APIs are never quick and easy to review. You can work on it and solicit feedback from other midders but it will probably sit open for a while

**Comment by ScalarVector1** [comment]
To me this builder-like syntax seems quite obtuse. Unlike recipes there aren't a large amount of unrelated/extra conditions you may need to apply each with their own parameters, I think a simpler solution would be to allow registration of simple conversions similar to tile merges, ala ```AddConversion(typeFrom, typeTo)```/```AddConversion(setFrom, typeTo)``` or something of a similar form. Special cases may better be handled in some sort of ``CanConvert/OnConvert`` hook for more consistency with the existing API and not requiring the use of delegation which may be confusing for newcomers.

Additionally, I think there is a need for more sane defaults than your example provides. Kill trees if invalid should likely be the default behavior.

Finally with the current implementation I question the need for an entire ModType here. It seems to essentially be only a ``SetStaticDefaults``, could this registration not simply be done in the SetStaticDefaults of a solution item or it's associated projectile? Is there a reason we need an entire ModType dedicated to conversions?

**Comment by BasicallyIAmFox** [comment]
> could this registration not simply be done in the SetStaticDefaults of a solution item or it's associated projectile?

hm, yeah, this can be somehow done in SetStaticDefaults of a solution item or projectile. Maybe something like:
`ConversionHandler.CreateConversion().Methods().Register(Type);`

> Special cases may better be handled in some sort of CanConvert/OnConvert hook for more consistency with the existing API and not requiring the use of delegation which may be confusing for newcomers.

imo this will result in repeatitive code (checking multiple times for whatever tile, etch) so i dont really want to have separate hooks :/

**Comment by BasicallyIAmFox** [comment]
@ScalarVector1

**Comment by Chicken-Bones** [comment]
Yeah, `ModConversion` is unnecessary. `Conversion` should be something like `Recipe` or `NPCShop`.

The builder syntax is fine, but not building the entire conversion in one go, there should be one call per tile/wall, more like `NPCShop`. No need for `From`, `To`, `RegisterTile` etc methods. Just `ConvertTile(from, to)` and `ConvertWall(from, to)`

**Comment by Nil256** [comment]
(I use google translation, so I maybe make some mistake of my English)
I can't agree this syntax, but I think `ModConversion` isn't always bad.
We become able to call like `WorldGen.Convert(i, j, ModContent.ConversionType<ModdedConversion>())` if tModLoader assign ID to each modded conversions.
Moreover mods can control other mods conversion if there is `GlobalConversion`.

```
public abstract class GlobalConversion : ...
{
    public virtual bool CanConvert(int i, int j, int type)
    {
        return true;
    }
}
```
`GlobalConversion.CanConvert()` should be called after checking the tile is in converting range in `WorldGen.Convert()`.
```
public abstract class ModConversion : ...
{
    public int Type => ...

    public virtual void Convert(int i, int j, int type)
    {
    }
}
```
`ModConversion .Convert()` should be called in default label of switch conversion type in `WorldGen.Convert()`.

In my opinion, the conversion is more complex than recipes and shops, so static register should not be used to the conversion.
(Ex: the conversion checks `TileID.Sets`/`WallID.Sets`, surrounding tiles / the conversion does not only change tiles but also just kill tiles. (Thorn) )
Reducing the code which should be write does not always turn into good.
Modded solutions have the potential.

Additionally, I think mods which need conversion is not many.
So I think it's no problem if the code which should be written is many a little.

There also is problem to this, I wonder if it is good to create new modtype for just 2 methods.
However we become not to need advanced/expert technology for cross mods at least if using this.

**Comment by JavidPack** [comment]
I believe https://github.com/tModLoader/tModLoader/pull/4566 might have covered all the features asked of in this issue.
